Sources/GRPC/GRPCServerRequestRoutingHandler.swift

```diff
@@ -0,0 +1,241 @@
+        context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
+        return
+      }
+
+      self.logger.debug("received request head, configuring pipeline")
+
+      // Buffer the request head; we'll replay it in the next handler when we're removed from the
+      // pipeline.
+      self.state = .configuring([requestPart])
+
+      // Configure the rest of the pipeline to serve the RPC.
+      let codec = callHandler.makeGRPCServerCodec()
+      context.pipeline.addHandlers([codec, callHandler], position: .after(self)).whenSuccess {
+        context.pipeline.removeHandler(self, promise: nil)
+      }
+
+    case .body, .end:
+      switch self.state {
+      case .notConfigured:
+        // We can reach this point if we're receiving messages for a method that isn't implemented,
+        // in which case we just drop the messages; our response should already be in-flight.
+        ()
+
+      case .configuring(var buffer):
+        // We received a message while the pipeline was being configured; hold on to it while we
+        // finish configuring the pipeline.
+        buffer.append(requestPart)
+        self.state = .configuring(buffer)
+      }
+    }
+  }
+
+  public func handlerRemoved(context: ChannelHandlerContext) {
+    switch self.state {
+    case .notConfigured:
+      ()
+
+    case .configuring(let messages):
+      for message in messages {
+        context.fireChannelRead(self.wrapInboundOut(message))
+      }
+    }
+  }
+
+  private func makeCallHandler(channel: Channel, requestHead: HTTPRequestHead) -> GRPCCallHandler? {
+    // URI format: "/package.Servicename/MethodName", resulting in the following components separated by a slash:
+    // - uriComponents[0]: empty
+    // - uriComponents[1]: service name (including the package name);
+    //     `CallHandlerProvider`s should provide the service name including the package name.
+    // - uriComponents[2]: method name.
+    self.logger.debug("making call handler", metadata: ["path": "\(requestHead.uri)"])
+    let uriComponents = requestHead.uri.components(separatedBy: "/")
+
+    var logger = self.logger
+    // Unset the channel handler: it shouldn't be used for downstream handlers.
+    logger[metadataKey: MetadataKey.channelHandler] = nil
+
+    let context = CallHandlerContext(errorDelegate: self.errorDelegate, logger: logger)
+
+    guard uriComponents.count >= 3 && uriComponents[0].isEmpty,
+      let providerForServiceName = servicesByName[uriComponents[1]],
+      let callHandler = providerForServiceName.handleMethod(uriComponents[2], callHandlerContext: context) else {
+        self.logger.notice("could not create handler", metadata: ["path": "\(requestHead.uri)"])
+        return nil
+    }
+    return callHandler
+  }
+
+  private func makeResponseHead(requestHead: HTTPRequestHead, status: GRPCStatus) -> HTTPResponseHead {
+    var headers: HTTPHeaders = [
+      "content-type": "application/grpc",
```

Comment:
should we use string constants for both the key and value?

Reply:
Good call.

Reply:
Did you check that these constants are being used everywhere now?